File: src/misc/code-types.ts
export type ProvinceCode = 'AB' | 'BC' | 'MB' | 'NB' | 'NL' | 'NS' | 'PE' | 'ON' | 'QC' | 'SK' | 'NT' | 'NU' | 'YT';
export type FederalCode = 'CA';
export type FederalName = 'CANADA';
export type ProvinceName =
    'ALBERTA'
    | 'BRITISH_COLUMBIA'
    | 'MANITOBA'
    | 'NEW_BRUNSWICK'
    | 'NEWFOUNDLAND'
    | 'NOVA_SCOTIA'
    | 'PRINCE_EDWARD_ISLAND'
    | 'ONTARIO'
    | 'QUEBEC'
    | 'SASKATCHEWAN'
    | 'NORTHWEST_TERRITORIES'
    | 'NUNAVUT'
    | 'YUKON';

export const PROVINCIAL_CODES: { [key in ProvinceName]: ProvinceCode } = {
    ALBERTA: 'AB',
    BRITISH_COLUMBIA: 'BC',
    MANITOBA: 'MB',
    NEW_BRUNSWICK: 'NB',
    NEWFOUNDLAND: 'NL',
    NOVA_SCOTIA: 'NS',
    PRINCE_EDWARD_ISLAND: 'PE',
    ONTARIO: 'ON',
    QUEBEC: 'QC',
    SASKATCHEWAN: 'SK',
    NORTHWEST_TERRITORIES: 'NT',
    NUNAVUT: 'NU',
    YUKON: 'YT',
};

export const FEDERAL_CODE: FederalCode = 'CA';
export type ByProvince<T> = { [key in ProvinceCode]: T };
export type ByJurisdiction<T> = { [key in ProvinceCode | FederalCode]: T };

File: src/commonMain/kotlin/com/equisoft/taxca/misc/CodeTypes.kt
package com.equisoft.taxca.misc

enum class Jurisdiction(val code: String, val jurisdictionName: String) {
    CA("CA", "CANADA"),
    AB("AB", "ALBERTA"),
    BC("BC", "BRITISH_COLUMBIA"),
    MB("MB", "MANITOBA"),
    NB("NB", "NEW_BRUNSWICK"),
    NL("NL", "NEWFOUNDLAND"),
    NS("NS", "NOVA_SCOTIA"),
    PE("PE", "PRINCE_EDWARD_ISLAND"),
    ON("ON", "ONTARIO"),
    QC("QC", "QUEBEC"),
    SK("SK", "SASKATCHEWAN"),
    NT("NT", "NORTHWEST_TERRITORIES"),
    NU("NU", "NUNAVUT"),
    YT("YT", "YUKON"),
    ;

    val isProvince: Boolean get() = this != CA

    companion object {
        val provinces: List<Jurisdiction> = entries.filter { it != CA }

        fun fromCode(code: String): Jurisdiction = entries.first { it.code == code }
    }
}
